sorrydb/agents/tactic_strategy.py

```diff
@@ -0,0 +1,527 @@
+                    if block.startswith("lean"):
+                        block = block[4:]
+                    extracted_code.append(block.strip())
+
+            # Use the first code block if any were found
+            if extracted_code:
+                response_text = extracted_code[0]
+
+        # Parse the state ID and tactic using regex
+        match = re.match(r"\[(\d+)\]\s*(.*)", response_text)
+
+        if not match:
+            logger.warning("Response doesn't match expected format: %s", response_text)
+            return None
+
+        state_id = int(match.group(1))
+        tactic = match.group(2).strip()
+
+        logger.info("Generated state ID: %d, tactic: %s", state_id, tactic)
+        return state_id, tactic
+
+    def _get_tactic_and_state(
+        self,
+        all_states: dict[int, list[str]],
+        search_history: list[str],
+        file_context: str = "",
+    ) -> tuple[int, str] | None:
+        """
+        Generates the next tactic to try and the proof state to apply it to.
+        Uses the selected strategy mode to determine how to generate the tactic.
+
+        Args:
+            all_states: Dictionary mapping proof state IDs to their lists of goal strings
+            search_history: Chronological history of the entire proof search
+            file_context: Context from the Lean file up to the sorry
+
+        Returns:
+            Tuple of (proof_state_id, tactic) or None if generation fails
+        """
+        # If we've had too many consecutive failures, give up
+        if self.consecutive_failures >= self.max_consecutive_failures:
+            logger.warning(
+                "Too many consecutive tactic failures (%d). Stopping.",
+                self.consecutive_failures,
+            )
+            return None
+
+        # Generate the tactic and state ID based on the strategy mode
+        result = self._generate_tactic(search_history, file_context)
+
+        if result is None:
+            return None
+
+        state_id, tactic = result
+
+        # Check if the state ID is valid
+        if state_id not in all_states:
+            logger.warning(
+                "Invalid state ID %d, available states: %s",
+                state_id,
+                list(all_states.keys()),
+            )
+            return None
+
+        return state_id, tactic
+
+    def prove_sorry(self, repo_path: Path, sorry: Sorry) -> str | None:
+        """Attempt to prove a sorry using the selected strategy mode.
+
+        Args:
+            repo_path: Path to the repository
+            sorry: Dictionary containing sorry information
+
+        Returns:
+            Proof string or None if no proof was found
+        """
+        # Load the file and render the prompt
+        loc = sorry.location
+        file_path = repo_path / loc.path
+        file_text = file_path.read_text()
+
+        # Extract the context up to the sorry line
+        context_lines = file_text.splitlines()[: loc.start_line]
```

Comment:
```suggestion
        context_lines = file_text.splitlines()[loc.start_line - max_context_lines : loc.start_line]
```
The current version potentially uses a huge amount of context; maybe it is better to truncate? We could make maxx_context_lines a CLI option (with default to e.g. 100).